drivers/gpio/gpio_bflb_bl60x_70x.c

```diff
@@ -0,0 +1,363 @@
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	uint32_t tmp;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+	tmp ^= mask;
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL32_OFFSET);
+
+	return 0;
+}
+
+static void gpio_bflb_port_interrupt_configure_mode(const struct device *dev, uint32_t pin,
+						    enum gpio_int_mode mode,
+						    enum gpio_int_trig trig)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	uint32_t tmp;
+	uint8_t trig_mode = GPIO_BFLB_TRIG_MODE_SYNC_LOW;
+
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_INT_MODE_SET1_OFFSET
+		+ ((pin / GPIO_BFLB_PIN_INT_PER_REG) << GPIO_BFLB_PIN_REG_SIZE_SHIFT));
+	/* clear modes */
+	tmp &= ~(GPIO_BFLB_PIN_INT_REG_MSK
+		<< ((pin % GPIO_BFLB_PIN_INT_PER_REG) * GPIO_BFLB_PIN_INT_REG_SIZE));
+
+	if ((trig & GPIO_INT_HIGH_1) != 0) {
+		trig_mode |= GPIO_BFLB_TRIG_MODE_SYNC_HIGH;
+	}
+
+	if ((mode & GPIO_INT_EDGE) == 0) {
+		trig_mode |= GPIO_BFLB_TRIG_MODE_SYNC_LEVEL;
+	}
+	tmp |= (trig_mode << ((pin % GPIO_BFLB_PIN_INT_PER_REG) * GPIO_BFLB_PIN_INT_REG_SIZE));
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_INT_MODE_SET1_OFFSET
+		+ ((pin / GPIO_BFLB_PIN_INT_PER_REG) << GPIO_BFLB_PIN_REG_SIZE_SHIFT));
+}
+
+static void gpio_bflb_pin_interrupt_clear(const struct device *dev, uint32_t mask)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+
+	sys_write32(mask, cfg->base_reg + GLB_GPIO_INT_CLR1_OFFSET);
+	clock_bflb_settle();
+	sys_write32(0x0, cfg->base_reg + GLB_GPIO_INT_CLR1_OFFSET);
+}
+
+static int gpio_bflb_pin_interrupt_configure(const struct device *dev,
+					     gpio_pin_t pin,
+					     enum gpio_int_mode mode,
+					     enum gpio_int_trig trig)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	uint32_t tmp;
+
+	/* Mask the interrupt. */
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_INT_MASK1_OFFSET);
+	tmp |= BIT(pin);
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_INT_MASK1_OFFSET);
+
+	gpio_bflb_port_interrupt_configure_mode(dev, pin, mode, trig);
+
+	if (mode != GPIO_INT_MODE_DISABLED) {
+		gpio_bflb_pin_interrupt_clear(dev, BIT(pin));
+		tmp = sys_read32(cfg->base_reg + GLB_GPIO_INT_MASK1_OFFSET);
+		tmp &= ~BIT(pin);
+		sys_write32(tmp, cfg->base_reg + GLB_GPIO_INT_MASK1_OFFSET);
+	} else {
+		tmp = sys_read32(cfg->base_reg + GLB_GPIO_INT_MASK1_OFFSET);
+		tmp |= BIT(pin);
+		sys_write32(tmp, cfg->base_reg + GLB_GPIO_INT_MASK1_OFFSET);
+	}
+	cfg->irq_enable_func(dev);
+
+	return 0;
+}
+
+static int gpio_bflb_config(const struct device *dev, gpio_pin_t pin,
+			   gpio_flags_t flags)
+{
+	const struct gpio_bflb_config * const cfg = dev->config;
+	uint8_t is_odd = 0;
+	uint32_t cfg_address;
+	uint32_t tmp;
+	uint32_t tmp_a;
+	uint32_t tmp_b;
+
+
+	/* Disable output anyway */
+	tmp = sys_read32(cfg->base_reg + GLB_GPIO_CFGCTL34_OFFSET + ((pin >> 5) << 2));
+	tmp &= ~(1U << (pin & 0x1f));
+	sys_write32(tmp, cfg->base_reg + GLB_GPIO_CFGCTL34_OFFSET + ((pin >> 5) << 2));
```

Comment:
Magic number and repetition

What is easier to read ?
tmp &= ~(1U << (pin & 0x1f));
or
tmp &= ~(BIT(pin & PIN_MASK));


This is very hard to understand what you want to do here `((pin >> 5) << 2)`. Even more because you are dropping the pin number information. So what you get when you shift 5 places to right ? No constant, no idea.

Looking in the datasheet, which is too much effort, should this be 
sys_clear32(BIT(pin | mask), cfg->base_reg + GLB_GPIO_CFGCTL34_OFFSET);

Reply:
This ((pin >> 5) << 2) for handling pin > 32, which seems to be completely useless here

Reply:
In this case, is this what you want to express?

(pin % 0x20) * 0x04